cni/pkg/plugin/kubernetes.go

```diff
@@ -67,47 +69,43 @@ func getK8sPodInfo(client *kubernetes.Clientset, podName, podNamespace string) (
 	}
 
 	pi := &PodInfo{
-		InitContainers:    make(map[string]struct{}),
-		Containers:        make([]string, len(pod.Spec.Containers)),
+		Containers:        sets.New[string](),
 		Labels:            pod.Labels,
 		Annotations:       pod.Annotations,
 		ProxyEnvironments: make(map[string]string),
 	}
-	for _, initContainer := range pod.Spec.InitContainers {
-		pi.InitContainers[initContainer.Name] = struct{}{}
-	}
-	for containerIdx, container := range pod.Spec.Containers {
-		log.Debugf("Inspecting pod %v/%v container %v", podNamespace, podName, container.Name)
-		pi.Containers[containerIdx] = container.Name
-
-		if container.Name == "istio-proxy" {
+	for _, c := range containers(pod) {
+		pi.Containers.Insert(c.Name)
+		if c.Name == "istio-proxy" {
 			// don't include ports from istio-proxy in the redirect ports
 			// Get proxy container env variable, and extract out ProxyConfig from it.
-			for _, e := range container.Env {
+			for _, e := range c.Env {
 				pi.ProxyEnvironments[e.Name] = e.Value
 			}
-
-			if container.SecurityContext != nil {
-				pi.ProxyUID = container.SecurityContext.RunAsUser
-				pi.ProxyGID = container.SecurityContext.RunAsGroup
+			if len(c.Args) >= 2 && c.Args[0] == "proxy" {
+				pi.ProxyType = c.Args[1]
+			}
+			if c.SecurityContext != nil {
+				pi.ProxyUID = c.SecurityContext.RunAsUser
+				pi.ProxyGID = c.SecurityContext.RunAsGroup
 			}
-
-			continue
 		}
 	}
 	log.Debugf("Pod %v/%v info: \n%+v", podNamespace, podName, pi)
 
 	return pi, nil
 }
 
+func containers(pod *v1.Pod) []v1.Container {
+	res := make([]v1.Container, 0, len(pod.Spec.Containers)+len(pod.Spec.InitContainers))
+	res = append(res, pod.Spec.InitContainers...)
+	res = append(res, pod.Spec.Containers...)
+	return res
+}
```

Comment:
I get that these are now combined for back compat with both KEP-style and non-KEP-style sidecar detection, but can we comment here stating that _eventually_ we want the KEP approach of init containers to be the sidecar default, and ultimately we should only be looking at init containers?

Reply:
even if it's the default we will support the old one. I won't say forever but with 4 old k8s versions supported, optimistically we are looking at 7 k8s versions of 2.5years min before we can remove the old mode.
I'll make the comments more clear